sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSchemaInferenceSuite.scala

```diff
@@ -0,0 +1,192 @@
+      storage = CatalogStorageFormat(
+        locationUri = Option(location),
+        inputFormat = Option("org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat"),
+        outputFormat = Option("org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat"),
+        serde = Option("org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"),
+        compressed = false,
+        properties = Map("serialization.format" -> "1")),
+      schema = schema,
+      provider = Option("hive"),
+      partitionColumnNames = partitionColumns,
+      properties = properties)
+  }
+
+  // Creates CatalogTablePartition instances for adding partitions of data to our test table.
+  private def hiveCatalogPartition(location: String, index: Int): CatalogTablePartition
+    = CatalogTablePartition(
+      spec = Map("partcol1" -> index.toString, "partcol2" -> index.toString),
+      storage = CatalogStorageFormat(
+        locationUri = Option(s"${location}/partCol1=$index/partCol2=$index/"),
+        inputFormat = Option("org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat"),
+        outputFormat = Option("org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat"),
+        serde = Option("org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"),
+        compressed = false,
+        properties = Map("serialization.format" -> "1")))
+
+  // Creates a case-sensitive external Hive table for testing schema inference options. Table
+  // will not have Spark-specific table properties set.
+  private def setupCaseSensitiveTable(
+      tableName: String,
+      dir: File): Unit = {
+    spark.range(NUM_RECORDS)
+      .selectExpr("id as fieldOne", "id as partCol1", "id as partCol2")
+      .write
+      .partitionBy("partCol1", "partCol2")
+      .mode("overwrite")
+      .parquet(dir.getAbsolutePath)
+
+
+    val client = externalCatalog.client
+
+    val catalogTable = hiveExternalCatalogTable(
+      tableName,
+      dir.getAbsolutePath,
+      lowercaseSchema,
+      Seq("partcol1", "partcol2"))
+    client.createTable(catalogTable, true)
+
+    val partitions = (0 until NUM_RECORDS).map(hiveCatalogPartition(dir.getAbsolutePath, _)).toSeq
+    client.createPartitions(DATABASE, tableName, partitions, true)
+  }
+
+  // Create a test table used for a single unit test, with data stored in the specified directory.
+  private def withTestTable(dir: File)(f: File => Unit): Unit = {
+    setupCaseSensitiveTable(TEST_TABLE_NAME, dir)
+    try f(dir) finally spark.sql(s"DROP TABLE IF EXISTS $TEST_TABLE_NAME")
+  }
+
+  private val inferenceKey = SQLConf.HIVE_CASE_SENSITIVE_INFERENCE.key
+
+  test("Schema should be inferred and written to table properties when INFER_AND_SAVE is " +
+    "specified") {
+    withSQLConf(inferenceKey -> HiveCaseSensitiveInferenceMode.INFER_AND_SAVE.toString) {
+      withTempDir { dir =>
+        withTestTable(dir) { dir =>
+          assert(spark.sql(FIELD_QUERY).count == NUM_RECORDS)
```

Comment:
Good call, I'll add a check for ```schemaPreservesCase```.

We'll have to use the underlying HiveClient to obtain the raw table in order to check the presence of ```DATASOURCE_SCHEMA_NUMPARTS``` instead of ```originalTable``` directly since HiveExternalCatalog [filters out any property starting with SPARK_SQL_PREFIX](https://github.com/apache/spark/blob/f48c5a57d6488d5598534ca5834e008504f464fe/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L665-L668).

I'm thinking of just adding these checks to the ```setupCaseSensitiveTable()``` method since we're essentially just asserting that our initial conditions are what we expect (table returned by catalog has schemaPreservesCase=false and the underlying table contains no Spark properties).